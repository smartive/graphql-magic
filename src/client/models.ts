export type ClientEntity = Record<string, unknown>;
