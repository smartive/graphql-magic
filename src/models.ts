import { DateTime } from 'luxon';
import { Field } from '.';
import type { Context } from './context';
import type { OrderBy } from './resolvers/arguments';
import type { Value } from './values';

export type RawModels = RawModel[];

export type RawModel = ScalarModel | EnumModel | RawEnumModel | InterfaceModel | ObjectModel | RawObjectModel;

type BaseModel = {
  name: string;
  plural?: string;
  description?: string;
};

export type ScalarModel = BaseModel & { type: 'scalar' };

export type EnumModel = BaseModel & { type: 'enum'; values: string[]; deleted?: true };

export type RawEnumModel = BaseModel & { type: 'raw-enum'; values: string[] };

export type InterfaceModel = BaseModel & { type: 'interface'; fields: ModelField[] };

export type RawObjectModel = BaseModel & {
  type: 'raw';
  fields: RawObjectField[];
};

export type Entity = Record<string, unknown> & { createdAt?: DateTime; deletedAt?: DateTime };

export type Action = 'create' | 'update' | 'delete' | 'restore';

export type MutationHook = (
  model: Model,
  action: Action,
  when: 'before' | 'after',
  data: { prev: Entity; input: Entity; normalizedInput: Entity; next: Entity },
  ctx: Context
) => Promise<void>;

export type ObjectModel = BaseModel & {
  type: 'object';
  interfaces?: string[];
  queriable?: boolean;
  listQueriable?: boolean;
  creatable?: boolean | { createdBy?: Partial<RelationField>; createdAt?: Partial<DateTimeField> };
  updatable?: boolean | { updatedBy?: Partial<RelationField>; updatedAt?: Partial<DateTimeField> };
  deletable?:
    | boolean
    | { deleted?: Partial<BooleanField>; deletedBy?: Partial<RelationField>; deletedAt?: Partial<DateTimeField> };
  displayField?: string;
  defaultOrderBy?: OrderBy;
  fields: ModelField[];

  // temporary fields for the generation of migrations
  deleted?: true;
  oldName?: string;
};

type BaseNumberType = {
  unit?: 'million';
  min?: number;
  max?: number;
};

type BaseField = Omit<Field, 'type'>;

type PrimitiveField =
  | { type: 'ID' }
  | { type: 'Boolean' }
  | {
      type: 'String';
      stringType?: 'email' | 'url' | 'phone';
      large?: true;
      maxLength?: number;
    }
  | {
      type: 'DateTime';
      dateTimeType?: 'year' | 'date' | 'datetime' | 'year_and_month';
      endOfDay?: boolean;
    }
  | ({
      type: 'Int';
      intType?: 'currency';
    } & BaseNumberType)
  | ({
      type: 'Float';
      floatType?: 'currency' | 'percentage';
      double?: boolean;
      precision?: number;
      scale?: number;
    } & BaseNumberType)
  | { type: 'Upload' };

type RawObjectField = BaseField & PrimitiveField;

export type ModelField = BaseField &
  (
    | PrimitiveField
    | { type: 'json'; typeName: string }
    | { type: 'enum'; typeName: string; possibleValues?: Value[] }
    | { type: 'raw'; typeName: string }
    | {
        type: 'relation';
        typeName: string;
        toOne?: boolean;
        reverse?: string;
        foreignKey?: string;
        onDelete?: 'cascade' | 'set-null';
      }
  ) & {
    primary?: boolean;
    unique?: boolean;
    filterable?:
      | boolean
      | {
          default?: Value;
        };
    searchable?: boolean;
    orderable?: boolean;
    comparable?: boolean;
    queriable?:
      | boolean
      | {
          roles?: string[];
        };
    creatable?:
      | boolean
      | {
          roles?: string[];
        };
    updatable?:
      | boolean
      | {
          roles?: string[];
        };
    generated?: boolean;
    // The tooltip is "hidden" behind an icon in the admin forms
    tooltip?: string;
    defaultValue?: string | number | ReadonlyArray<string> | undefined;
    // If true the field must be filled within forms but can be null in the database
    required?: boolean;
    indent?: boolean;
    // If true the field is hidden in the admin interface
    hidden?: boolean;

    // temporary fields for the generation of migrations
    deleted?: true;
    oldName?: string;
  };

export type IDField = Extract<ModelField, { type: 'ID' }>;
export type BooleanField = Extract<ModelField, { type: 'Boolean' }>;
export type StringField = Extract<ModelField, { type: 'String' }>;
export type DateTimeField = Extract<ModelField, { type: 'DateTime' }>;
export type IntField = Extract<ModelField, { type: 'Int' }>;
export type FloatField = Extract<ModelField, { type: 'Float' }>;
export type JsonField = Extract<ModelField, { type: 'json' }>;
export type EnumField = Extract<ModelField, { type: 'enum' }>;
export type RawField = Extract<ModelField, { type: 'raw' }>;
export type RelationField = Extract<ModelField, { type: 'relation' }>;

export type Models = Model[];

export type Model = ObjectModel & {
  fieldsByName: Record<string, ModelField>;
  relations: Relation[];
  relationsByName: Record<string, Relation>;
  reverseRelations: ReverseRelation[];
  reverseRelationsByName: Record<string, ReverseRelation>;
};

export type Relation = {
  field: RelationField;
  model: Model;
  reverseRelation: ReverseRelation;
};

export type ReverseRelation = RelationField & {
  model: Model;
  field: RelationField;
  fieldModel: Model;
};
