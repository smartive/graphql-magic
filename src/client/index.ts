// created from 'create-ts-index'

export * from './gql';
export * from './queries';
