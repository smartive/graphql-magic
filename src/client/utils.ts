import upperCase from 'lodash/upperCase';

export const constantCase = (str: string) => upperCase(str).replace(/ /g, '_');
