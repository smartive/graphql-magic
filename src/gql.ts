// This tag does nothing (just generates a string) - it is here for the tooling (syntax highlighting, formatting and type generation)
export const gql = (chunks: TemplateStringsArray, ...variables: (string | number | boolean)[]): string => {
  return chunks.reduce(
    (accumulator, chunk, index) => `${accumulator}${chunk}${index in variables ? variables[index] : ''}`,
    ''
  );
};
