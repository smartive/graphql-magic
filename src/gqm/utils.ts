import { SourceFile, SyntaxKind, SyntaxList } from 'ts-morph';

export const findDeclarationInFile = (sourceFile: SourceFile, name: string) => {
  const syntaxList = sourceFile.getChildrenOfKind(SyntaxKind.SyntaxList)[0];
  if (!syntaxList) {
    throw new Error('No SyntaxList');
  }
  const declaration = findDeclaration(syntaxList, name);
  if (!declaration) {
    throw new Error('No rawModels declaration');
  }
  return declaration;
};

const findDeclaration = (syntaxList: SyntaxList, name: string) => {
  for (const variableStatement of syntaxList.getChildrenOfKind(SyntaxKind.VariableStatement)) {
    for (const declaration of variableStatement.getDeclarationList().getDeclarations()) {
      if (declaration.getName() === name) {
        return declaration;
      }
    }
  }
};
