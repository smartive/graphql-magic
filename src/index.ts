// created from 'create-ts-index'

export * from './api';
export * from './client';
export * from './db';
export * from './migrations';
export * from './models';
export * from './permissions';
export * from './resolvers';
export * from './schema';
export * from './utils';
export * from './context';
export * from './errors';
export * from './values';
