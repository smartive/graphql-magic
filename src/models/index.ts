// created from 'create-ts-index'

export * from './model-definitions';
export * from './models';
export * from './mutation-hook';
export * from './utils';
