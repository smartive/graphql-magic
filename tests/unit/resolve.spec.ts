import knex from 'knex';
import { DateTime } from 'luxon';
import { execute } from '../../src';
import { gql } from '../../src/client/gql';
import { getResolvers } from '../../src/resolvers';
import { models, permissions, rawModels } from '../utils/models';

const test = async (operationName: string, query: string, variables: object, responses: unknown[]) => {
  const knexInstance = knex({
    client: 'postgresql',
  });

  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const mockKnex = require('mock-knex');
  mockKnex.mock(knexInstance);
  const tracker = mockKnex.getTracker();

  tracker.install();
  tracker.on('query', function someFunction(query, step) {
    expect(query.sql).toMatchSnapshot('query');
    query.response(responses[step - 1]);
  });

  const user = await knexInstance('User').where({ id: 1 }).first();
  const result = await execute({
    knex: knexInstance,
    locale: 'en',
    locales: ['en'],
    user,
    rawModels,
    models,
    permissions,
    now: DateTime.fromISO('2020-01-01T00:00:00.000Z'),
    body: { operationName, query, variables }
  });

  expect(result).toMatchSnapshot();

  tracker.uninstall();
};

describe('resolvers', () => {
  it('are generated correctly', () => {
    expect(getResolvers(models)).toMatchSnapshot();
  });

  it('resolve lists, many-to-one and one-to-many queries', async () => {
    await test(
      'SomeQuery',
      gql`
        query SomeQuery {
          manyObjects(where: { another: { id: "bar" } }, orderBy: [{ xyz: DESC }]) {
            id
            field
            another {
              id
              manyObjects(where: { id: "foo" }) {
                id
                field
              }
            }
          }
        }
      `,
      {},
      [
        { id: 1, role: 'ADMIN' },
        [
          {
            SO__ID: 'foo',
            SO__id: 'foo',
            SO__field: 'foo',
            SO__a__ID: 'bar',
            SO__a__id: 'bar',
          },
        ],
        [
          {
            SO__a__mO__ID: 'foo',
            SO__a__mO__id: 'foo',
            SO__a__mO__field: 'foo',
            SO__a__mO__anotherId: 'bar',
          },
        ],
      ]
    );
  });

  it('resolve single query', async () => {
    await test(
      'SomeQuery',
      gql`
        query SomeQuery {
          someObject(where: { id: "foo" }) {
            id
            field
          }
        }
      `,
      {},
      [
        { id: 1, role: 'ADMIN' },
        [
          {
            SO__ID: 'foo',
            SO__id: 'foo',
            SO__field: 'foo',
          },
        ],
      ]
    );
  });
});
