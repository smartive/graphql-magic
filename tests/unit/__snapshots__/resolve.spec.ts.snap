// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolvers are generated correctly 1`] = `
{
  "Mutation": {
    "createSomeObject": [Function],
    "deleteAnotherObject": [Function],
    "deleteSomeObject": [Function],
    "restoreAnotherObject": [Function],
    "restoreSomeObject": [Function],
    "updateSomeObject": [Function],
  },
  "Query": {
    "anotherObjects": [Function],
    "manyObjects": [Function],
    "me": [Function],
    "someObject": [Function],
  },
}
`;

exports[`resolvers resolve lists, many-to-one and one-to-many queries 1`] = `
{
  "data": {
    "manyObjects": [
      {
        "another": {
          "id": "bar",
          "manyObjects": [
            {
              "field": "foo",
              "id": "foo",
            },
          ],
        },
        "field": "foo",
        "id": "foo",
      },
    ],
  },
}
`;

exports[`resolvers resolve lists, many-to-one and one-to-many queries: query 1`] = `"select * from "User" where "id" = $1 limit $2"`;

exports[`resolvers resolve lists, many-to-one and one-to-many queries: query 2`] = `"select "SO"."id" as "SO__ID", "SO"."id" as "SO__id", "SO"."field" as "SO__field", "SO__a"."id" as "SO__a__ID", "SO__a"."id" as "SO__a__id" from "SomeObject" as "SO" left join "AnotherObject" as "SO__W__a" on "SO"."anotherId" = "SO__W__a"."id" left join "AnotherObject" as "SO__a" on "SO"."anotherId" = "SO__a"."id" where "SO__W__a"."id" in ($1) and "SO"."deleted" = $2 order by "SO"."xyz" DESC"`;

exports[`resolvers resolve lists, many-to-one and one-to-many queries: query 3`] = `"select "SO__a__mO"."id" as "SO__a__mO__ID", "SO__a__mO"."id" as "SO__a__mO__id", "SO__a__mO"."field" as "SO__a__mO__field", "SO__a__mO"."anotherId" as "SO__a__mO__anotherId" from "SomeObject" as "SO__a__mO" where "SO__a__mO"."id" in ($1) and "SO__a__mO"."deleted" = $2 and "SO__a__mO"."anotherId" = $3 order by "SO__a__mO"."createdAt" DESC"`;

exports[`resolvers resolve single query 1`] = `
{
  "data": {
    "someObject": {
      "field": "foo",
      "id": "foo",
    },
  },
}
`;

exports[`resolvers resolve single query: query 1`] = `"select * from "User" where "id" = $1 limit $2"`;

exports[`resolvers resolve single query: query 2`] = `"select "SO"."id" as "SO__ID", "SO"."id" as "SO__id", "SO"."field" as "SO__field" from "SomeObject" as "SO" where "SO"."id" = $1 and "SO"."deleted" = $2 order by "SO"."createdAt" DESC limit $3"`;
